### PR TITLE
FEATURE: [max] new  method for depth buffer to set snapshot

### PR DESCRIPTION
**Pull request by kbearXD**

### CODE REVIEW

**Review by ycdesu — APPROVED** [review]

**Review by c9s — APPROVED** [review]

File: pkg/depth/buffer.go [hunk 1]
```diff
@@ -75,6 +75,34 @@ func (b *Buffer) Reset() {
...
+	}
+
+	b.mu.Lock()
+
+	if b.finalUpdateID >= finalUpdateID {
+		return nil
+	}
+
+	// clean the buffer since we already set the snapshot
+	b.buffer = nil
```
Comment (c9s) [inline pkg/depth/buffer.go:91]: i guess we can still keep the updates?

File: pkg/depth/buffer.go [hunk 2]
```diff
@@ -75,6 +75,34 @@ func (b *Buffer) Reset() {
...
+func (b *Buffer) SetSnapshot(snapshot types.SliceOrderBook, firstUpdateID int64, finalArgs ...int64) error {
+	finalUpdateID := firstUpdateID
+	if len(finalArgs) > 0 {
+		finalUpdateID = finalArgs[0]
+	}
+
+	b.mu.Lock()
+
+	if b.finalUpdateID >= finalUpdateID {
+		return nil
```
Comment (c9s) [inline pkg/depth/buffer.go:87]: unlock before return?

File: pkg/exchange/max/stream.go [hunk 3]
```diff
@@ -255,13 +255,26 @@ func (s *Stream) handleBookEvent(ex *Exchange) func(e max.BookEvent) {
...
 		}
 
-		if err := f.AddUpdate(types.SliceOrderBook{
-			Symbol: symbol,
-			Time:   e.Time(),
-			Bids:   e.Bids,
-			Asks:   e.Asks,
-		}, e.FirstUpdateID, e.LastUpdateID); err != nil {
-			log.WithError(err).Errorf("found missing %s update event", e.Market)
+		if e.Event == "snapshot" {
```
Comment (c9s) [inline pkg/exchange/max/stream.go:258]: do we have a constant for this? in the maxapi/v3 package.

Comment (kbearXD) [reply pkg/exchange/max/stream.go:258]: No, we bypass this to book event.

File: pkg/depth/buffer.go [hunk 4]
```diff
@@ -75,6 +75,34 @@ func (b *Buffer) Reset() {
 	b.mu.Unlock()
 }
 
+func (b *Buffer) SetSnapshot(snapshot types.SliceOrderBook, firstUpdateID int64, finalArgs ...int64) error {
```
Comment (c9s) [inline pkg/depth/buffer.go:78]: make it private method?

**Pull request merged**